src/main/java/com/conveyal/r5/rastercost/ElevationCostField.java

```diff
@@ -0,0 +1,177 @@
+package com.conveyal.r5.rastercost;
+
+import com.conveyal.r5.streets.EdgeStore;
+import gnu.trove.list.TFloatList;
+import gnu.trove.list.TShortList;
+import gnu.trove.list.array.TFloatArrayList;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Serializable;
+import java.util.List;
+
+import static com.conveyal.r5.rastercost.ElevationLoader.ELEVATION_SAMPLE_SPACING_METERS;
+
+/**
+ * This represents the changes in elevation along every edge in the network, as well as the derived additional costs
+ * for traversing those elevation changes.
+ *
+ * Inspecting the resulting edge elevation profiles reveals that a large number of them are the same elevation over the
+ * entire length of the edge, or very smooth linear changes. These can be simplified and compressed in several ways:
+ * 1. Where both end vertices and every sample point are at the same elevation, the sample points can simply be removed.
+ * 2. Where the sample points fall roughly on a straight line between the two end vertices the sample points are not
+ * needed. Item 1 is a special case of item 2.
+ * 3. ...
+ * 4. If after all these simplifications less than half of the edges have an explicit elevation profile, we may want to
+ * store them in a map instead of an array for better cache efficiency.
+ *
+ * The underlying per-edge storage mechanism is not shared with the tree shade costs because they have different
+ * characteristics. The tree shade is boolean and can be stored as only the distances along the edge where we enter or
+ * exit shade.
+ *
+ * However, storing alternating (distance, elevation change) in a packed array of shorts or bytes would be much easier
+ * to deal with when splitting edges.
+ *
+ * In this case tree shade might be able to share the storage model. The value change elements would all be +/- 1 which
+ * would waste some space, but the code reuse might be worth it. And of course this would also allow greyscale shade
+ * instead of binary shade inputs.
+ */
+public class ElevationCostField implements CostField, Serializable {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ElevationCostField.class);
+
+    public static final double DECIMETERS_PER_METER = 10;
```

Comment:
This is actually independent of `inputScale`. It's an attempt to keep the storage of elevation data down to a reasonable size since we are keeping a profile for every edge just to facilitate splitting them. Note that vertexElevationsDecimeters is a TShortList and elevationProfilesDecimeters contains short arrays, so all our elevation measurements take half as much space as 32 bit ints or a quarter as much as double precision floats. 2^15/10 meters is enough to store all elevations outside very high mountain passes. This may be considered premature or misguided optimization, its effectiveness should be evaluated.

Reply:
I replicated my above comment in the code.